lib/render/RayCaster.cpp

```diff
@@ -541,94 +687,111 @@ int RayCaster::_paintGL( bool fast )
-    cameraUser[1]         = InversedMV[ 13 ];
-    cameraUser[2]         = InversedMV[ 14 ];
+    cameraUser[0]         = InversedMV[3][0];
+    cameraUser[1]         = InversedMV[3][1];
+    cameraUser[2]         = InversedMV[3][2];
     std::vector<size_t> cameraCellIndices;      // camera position in which cell?
-    StructuredGrid*  grid = _userCoordinates.GetCurrentGrid( params, _dataMgr );
+    StructuredGrid*  grid = nullptr;
+    if( _userCoordinates.GetCurrentGrid( params, _dataMgr, &grid ) != 0 )
+    {
+        MyBase::SetErrMsg( "Failed to retrieve a StructuredGrid" );
+        return 1;
+    }
     bool insideACell      =  grid->GetIndicesCell( cameraUser, cameraCellIndices ); 
 
     if( insideACell )
     {
-        GLfloat MVP[16],    InversedMVP[16];
-        _getMVPMatrix( MVP );
-        _mesa_invert_matrix_general( InversedMVP, MVP );
-
-        float topLeftNDC[4]     = {-1.0f,  1.0f, -0.999f, 1.0f};
-        float bottomLeftNDC[4]  = {-1.0f, -1.0f, -0.999f, 1.0f};
-        float topRightNDC[4]    = { 1.0f,  1.0f, -0.999f, 1.0f};
-        float bottomRightNDC[4] = { 1.0f, -1.0f, -0.999f, 1.0f};
-        float near[16];
-        _matMultiVec( InversedMVP, topLeftNDC,     near     );
-        _matMultiVec( InversedMVP, bottomLeftNDC,  near + 4 );
-        _matMultiVec( InversedMVP, topRightNDC,    near + 8 );
-        _matMultiVec( InversedMVP, bottomRightNDC, near + 12);
+        glm::mat4 MVP         = mm->GetModelViewProjectionMatrix();
+        glm::mat4 InversedMVP = glm::inverse( MVP );
+        glm::vec4 topLeftNDC    ( -1.0f,  1.0f, -0.9999f, 1.0f );
+        glm::vec4 bottomLeftNDC ( -1.0f, -1.0f, -0.9999f, 1.0f );
+        glm::vec4 topRightNDC   ( 1.0f,  1.0f, -0.9999f, 1.0f );
+        glm::vec4 bottomRightNDC( 1.0f, -1.0f, -0.9999f, 1.0f );
+        glm::vec4 near[4];
+        near[0] = InversedMVP * topLeftNDC;
+        near[1] = InversedMVP * bottomLeftNDC;
+        near[2] = InversedMVP * topRightNDC;
+        near[3] = InversedMVP * bottomRightNDC;
         for( int i = 0; i < 4; i++ )
         {
-            _userCoordinates.nearCoords[ i*3   ] = near[ i*4   ] / near[ i*4+3 ];
-            _userCoordinates.nearCoords[ i*3+1 ] = near[ i*4+1 ] / near[ i*4+3 ];
-            _userCoordinates.nearCoords[ i*3+2 ] = near[ i*4+2 ] / near[ i*4+3 ];
+            near[i] /= near[i].w;
+            std::memcpy( _userCoordinates.nearCoords + i * 3, 
+                         glm::value_ptr(near[i]), 3 * sizeof(float) );
         }
     }
 
-    _drawVolumeFaces( 2, insideACell );    // 2nd pass, render front facing polygons
+    // 2nd pass, render front facing polygons
+    _drawVolumeFaces( 2, castingMode, insideACell );    
 
     glBindFramebuffer( GL_FRAMEBUFFER, 0 );
     glViewport( 0, 0, _currentViewport[2], _currentViewport[3] );
 
-    _drawVolumeFaces( 3, insideACell, ModelView, InversedMV, fast );  // 3rd pass, perform ray casting
+    // 3rd pass, perform ray casting
+    if(  castingMode == 1 )
```

Comment:
Could you make castingMode an enum or at least #define some constants? It is compared in multiple places with unlabeled integers.

Reply:
With only 2 casting modes, I feel either enum or macros are over killing. Thus I reject this change request.

Reply:
I disagree. I don't think I should have to go to the header file of a different class to find out what rendering mode 1 vs 2 is.

Reply:
Please make an enum or a boolean flag. Something with a name indicating what casting mode means

Reply:
Fixed in 4d41f667